core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.builtins.nullable
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+
+/*
+ * By default, we use ClassValue-based caches to avoid classloader leaks,
+ * but ClassValue is not available on Android, thus we attempt to check it dynamically
+ * and fallback to ConcurrentHashMap-based cache.
+ */
+private val useClassValue = runCatching {
+    Class.forName("java.lang.ClassValue")
+}.map { true }.getOrDefault(false)
+
+/**
+ * Creates a **strongly referenced** cache of values associated with [Class].
+ * Serializers are computed using provided [factory] function.
+ *
+ * `null` values are not supported, though there aren't any technical limitations.
+ */
+internal actual fun createCache(factory: (KClass<*>) -> KSerializer<Any>?): SerializerCache {
+    return if (useClassValue) ClassValueCache(factory) else ConcurrentHashMapCache(factory)
+}
+
+/**
+ * Creates a **strongly referenced** cache of values associated with [Class].
+ * Serializers are computed using provided [factory] function.
+ *
+ * `null` values are not supported, though there aren't any technical limitations.
+ */
+internal actual fun createParametrizedCache(factory: (KClass<Any>, List<KType>) -> KSerializer<Any>?): ParametrizedSerializerCache {
+    return if (useClassValue) ClassValueParametrizedCache(factory) else ConcurrentHashMapParametrizedCache(factory)
+}
+
+@SuppressAnimalSniffer
+private class ClassValueCache(private val compute: (KClass<*>) -> KSerializer<Any>?) : SerializerCache {
```

Comment:
Have we also considered a strategy for limiting max size/cleaning these caches?

While serialization is generally used often throughout the application, it is not a rare case when it is also used on the startup for e.g. configuration reading. In this case, `decodeFromString<MyConfiguration>()` creates a cache entry that is never used again and is never removed.

Reply:
That was actually my call to avoid limits for now.

In our initial mental model, these caches are nothing but static constant overhead.

E.g. if one uses `decodeFromString<MyConfiguration>()`, they opt-in into **some** footprint -- `companion object`, a few more loaded classes, few kilobytes of descriptors and other serialization-related stuff reachable from this companion and so on. This cache is an epsilon for this footprint, but instead of being reachable from a companion, it is reachable from here.

Nothing prevents us from introducing public `clearCaches` if there is need to, but for now it doesn't seem to be the case

Reply:
Given that configuration is likely one-two classes, I don't mind that static overhead. Nullability problem (see other comment) is worse IMO